lib/SIL/SILModule.cpp

```diff
@@ -762,3 +763,22 @@ bool SILModule::isOptimizedOnoneSupportModule() const {
   return getOptions().Optimization >= SILOptions::SILOptMode::Optimize &&
          isOnoneSupportModule();
 }
+
+void SILModule::setSerializeSILAction(SILModule::ActionCallback Action) {
+  assert(!SerializeSILAction && "Serialization action can be set only once");
+  SerializeSILAction = Action;
+}
+
+SILModule::ActionCallback SILModule::getSerializeSILAction() const {
+  return SerializeSILAction;
+}
+
+void SILModule::serialize() {
+  assert(SerializeSILAction && "Serialization action should be set");
+  // Make sure that the module is serialized only once.
+  if (isSerialized())
```

Comment:
I really don't like this; it means you can call serialize and silently not get the latest copy of the module. Can we move all serialization into the SIL module, or have the caller perform the check at least?

Reply:
Yeah, I think it would make more sense to split up the optimizer pipeline into two stages, and have the frontend tool explicitly serialize everything in the middle. Maybe @jrose-apple you should file a bug to clean this up later? I see @swiftix already merged a different version of this PR.